### PR TITLE
(BOLT-1035) Add get_resources function

### PR DESCRIPTION
**Pull request by MikaelSmith**
Adds a `get_resources` function that can be run against targets with Puppet Agent installed to query the state of resources on those nodes.

### CODE REVIEW

**Comment by donoghuc** [comment]
I'm sure that this is unrelated to this PR, but how does one specify a "fully qualified" path on windows? I am trying to run something like 
```
plan basic::resources(TargetSpec $nodes) {
  return $nodes.get_resources(['File[C:/Users/Administrator/Desktop/hi.txt]'])
}
```
And keep getting an error like:
```
{
  "kind": "bolt/run-failure",
  "msg": "Plan aborted: run_task 'apply_helpers::query_resources' failed on 1 nodes",
  "details": {
    "action": "run_task",
    "object": "apply_helpers::query_resources",
    "result_set": [
      {
        "node": "mhsweokn9guwqmk.delivery.puppetlabs.net",
        "status": "failure",
        "result": {
          "_output": "",
          "_error": {
            "kind": "puppetlabs.tasks/task-error",
            "issue_code": "TASK_ERROR",
            "msg": "The task failed with exit code 1:\n/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/errors.rb:157:in `fail': Parameter path failed on File[C:/Users/Administrator/Desktop/hi.txt]: File paths must be fully qualified, not 'C:/Users/Administrator/Desktop/hi.txt' (Puppet::ResourceError)\n\tfrom /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type/file.rb:56:in `block (3 levels) in <top (required)>'\n\tfrom /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/parameter.rb:461:in `validate'\n\tfrom /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/parameter.rb:496:in `value='\n\tfrom /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type.rb:674:in `[]='\n\tfrom /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type.rb:2469:in `set_name'\n\tfrom /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type.rb:2392:in `initialize'\n\tfrom /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type/file.rb:490:in `initialize'\n\tfrom /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/indirector/resource/ral.rb:15:in `new'\n\tfrom /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/indirector/resource/ral.rb:15:in `find'\n\tfrom /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/indirector/indirection.rb:194:in `find'\n\tfrom /tmp/65453fc8-39c3-4fdb-ad6c-fa5dbf8310c3/query_resources.rb:41:in `block (2 levels) in <main>'\n\tfrom /tmp/65453fc8-39c3-4fdb-ad6c-fa5dbf8310c3/query_resources.rb:39:in `each'\n\tfrom /tmp/65453fc8-39c3-4fdb-ad6c-fa5dbf8310c3/query_resources.rb:39:in `flat_map'\n\tfrom /tmp/65453fc8-39c3-4fdb-ad6c-fa5dbf8310c3/query_resources.rb:39:in `block in <main>'\n\tfrom /opt/puppetlabs/puppet/lib/ruby/2.5.0/tmpdir.rb:89:in `mktmpdir'\n\tfrom /tmp/65453fc8-39c3-4fdb-ad6c-fa5dbf8310c3/query_resources.rb:18:in `<main>'\n",
            "details": {
              "exit_code": 1
            }
          }
        }
      }
    ]
  }
}

```

**Comment by MikaelSmith** [comment]
Umm. Good question.

**Comment by MikaelSmith** [comment]
Was that a Windows target? https://puppet.com/docs/puppet/6.0/types/file.html#file-attribute-path seems to say it should work. The result_set pertains to a target, so I don't think it's the compilation erroring.

**Comment by MikaelSmith** [comment]
I think Puppet's check there is platform-specific. So hard to write a platform-agnostic file resource query. Checking files in general are a bad example, just easy to write and understand.

File: spec/integration/apply_spec.rb [hunk 1]
```diff
@@ -105,6 +105,18 @@ def agent_version_inventory
         end
       end
 
+      it 'gets resources' do
+        with_tempfile_containing('inventory', YAML.dump(agent_version_inventory), '.yaml') do |inv|
```
[the remark this answers is not in the document]
Comment (MikaelSmith) [reply spec/integration/apply_spec.rb]: I'd rather not add Windows testing. We test 5 and 6 for API compatibility. Cross-platform behavior of Puppet I expect to have tested with Puppet, and it'll behave however it behaves.

File: bolt-modules/boltlib/lib/puppet/functions/get_resources.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    param 'Variant[String, Resource, Array[Variant[String, Resource]]]', :resources
+  end
+
+  def script_compiler
+    @script_compiler ||= Puppet::Pal::ScriptCompiler.new(closure_scope.compiler)
+  end
+
+  def run_task(executor, targets, name, args = {})
+    tasksig = script_compiler.task_signature(name)
+    raise Bolt::Error.new("#{name} could not be found", 'bolt/apply-prep') unless tasksig
```
Comment (adreyer) [inline bolt-modules/boltlib/lib/puppet/functions/get_resources.rb]: this doesn't seem like the right error kind

File: bolt-modules/boltlib/lib/puppet/functions/get_resources.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    targets = inventory.get_targets(target_spec)
+
+    executor.log_action('gather resources', targets) do
+      executor.without_default_logging do
+        # Gather facts, including custom facts
+        plugins = applicator.build_plugin_tarball do |mod|
+          search_dirs = []
+          search_dirs << mod.plugins if mod.plugins?
+          search_dirs << mod.pluginfacts if mod.pluginfacts?
```
Comment (adreyer) [inline bolt-modules/boltlib/lib/puppet/functions/get_resources.rb]: Are pluginfacts used in provider tests?

File: bolt-modules/boltlib/lib/puppet/functions/get_resources.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+# Requires the Puppet Agent be installed on the target, which can be accomplished with apply_prep
+# or by directly running the puppet_agent::install task.
+Puppet::Functions.create_function(:get_resources) do
+  # @param targets A pattern or array of patterns identifying a set of targets.
+  # @param resources A resource type or instance, or an array of such.
+  # @example Collect resource states for packages and a file
+  #   get_resources('target1,target2', [Package, File[/etc/puppetlabs]])
+  dispatch :get_resources do
+    param 'Boltlib::TargetSpec', :targets
+    param 'Variant[String, Resource, Array[Variant[String, Resource]]]', :resources
```
Comment (adreyer) [inline bolt-modules/boltlib/lib/puppet/functions/get_resources.rb]: We should probably have a regex for the strings. perhaps inside the function so we can provide a better error.

File: bolt-modules/boltlib/lib/puppet/functions/get_resources.rb (hunk 3 above)
Comment (MikaelSmith) [reply bolt-modules/boltlib/lib/puppet/functions/get_resources.rb]: Yes.

**Comment by MikaelSmith** [comment]
Updated.

File: spec/integration/apply_spec.rb [hunk 5]
```diff
@@ -105,6 +105,18 @@ def agent_version_inventory
...
+          results = run_cli_json(%W[plan run basic::resources --nodes agent_targets
+                                    --modulepath #{modulepath} --inventoryfile #{inv.path}])
+          results.each do |result|
+            expect(result['status']).to eq('success')
+            resources = result['result']['resources']
+            expect(resources.map { |r| r['type'] }.uniq).to eq(%w[User File])
```
Comment (adreyer) [inline spec/integration/apply_spec.rb]: Should we make sure File['/tmp'] is returned and is the only File resource and that at least the bolt user is returned?

Comment (MikaelSmith) [reply spec/integration/apply_spec.rb]: Done.

File: bolt-modules/boltlib/spec/functions/get_resources_spec.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    Puppet.override(bolt_executor: executor, bolt_inventory: inventory, apply_executor: applicator) do
+      example.run
+    end
+  end
+
+  context 'with targets' do
+    let(:hostnames) { %w[a.b.com winrm://x.y.com pcp://foo] }
+    let(:targets) { hostnames.map { |h| Bolt::Target.new(h) } }
+    let(:unknown_targets) { targets.reject { |target| target.protocol == 'pcp' } }
```
Comment (lucywyman) [inline bolt-modules/boltlib/spec/functions/get_resources_spec.rb]: It looks like this doesn't get used

File: lib/bolt_spec/plans.rb [hunk 7]
```diff
@@ -202,7 +202,7 @@ def run_plan(name, params)
     end
 
     def allow_apply_prep
-      allow_task('apply_helpers::custom_facts')
+      allow_task('apply_helpers::custom_facts').be_called_times(100)
```
Comment (lucywyman) [inline lib/bolt_spec/plans.rb]: Why the 100 limit?

File: bolt-modules/boltlib/spec/functions/get_resources_spec.rb (hunk 6 above)
Comment (MikaelSmith) [reply bolt-modules/boltlib/spec/functions/get_resources_spec.rb]: That appears to be true.

File: lib/bolt_spec/plans.rb (hunk 7 above)
Comment (MikaelSmith) [reply lib/bolt_spec/plans.rb]: It was a hack, 100 seemed unlikely to be reached for apply_prep or get_resources in a plan.

Comment (MikaelSmith) [reply lib/bolt_spec/plans.rb]: Probably needs to be reworked to not enforce a limit unless explicitly specified. Which might not be hard now that I think about it.

**Pull request merged**